### PR TITLE
Angular: remove modal from entity crud

### LINKED ISSUE #7352: Remove Popup from CRUD navigation
Hi, i am using JHipster and think current navigation make JHipster code complex and management of code is very hard ,one of big problem in business application is change management and upgrade to new feature . 
I suggest change navigation to this way.
When user click on view or edit button .route change and component render in default outlet instead of open in popup .
When user click on delete button .confirm dialog open and after user select ok .entity remove and grid reload;
 This gift show this navigation .

mar-24-2018 19-36-34

In this navigation .all pop component and service remove and create on resolver for load data in resolver;

Current navigation 

Disadvantage:
1. Complex code 
2. More Code 
3. High cohesion: change propert outside of component using componentInstance in (entity-detail.component.ts and entity-dialog.component.ts) 
4. Popup dose not good for a lot of  input and complex form 

what is opinion of jhipster team about this navigation and advantage and disadvantage of this scenario?

### PR DESCRIPTION
**Pull request by azizkhani**
Fix #7352
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by deepu105** [comment]
i'll work on this and merge it this weekend

**Comment by pascalgrimaud** [comment]
Just amazing for a 1st contribution, congrats @azizkhani

**Comment by azizkhani** [comment]
I wished to be a contributer of jhipster, Jhipster is one of best project i have ever seen in last 10 years and learn too many thinks, i hope i can do more issue

**Comment by deepu105** [comment]
you are welcome @azizkhani and congrats for your PR, I will review and merge it this weekend

**Comment by jdubois** [comment]
Congrats! That's amazing, I didn't think we would get this for JHipster 5!

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs [hunk 1]
```diff
@@ -22,7 +22,7 @@ _%>
 <form name="editForm" role="form" novalidate (ngSubmit)="save()" #editForm="ngForm">
 
     <div>
-        <h2  id="my<%= entityClass %>Label" jhiTranslate="<%= keyPrefix %>home.createOrEditLabel">Create or edit a <%= entityClassHumanized %></h2>
+        <h4  id="my<%= entityClass %>Label" jhiTranslate="<%= keyPrefix %>home.createOrEditLabel">Create or edit a <%= entityClassHumanized %></h4>
```
Comment (gmarziou) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: Why this change?

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: Failed: No element found using locator: By(css selector, .jh-create-entity) in travis

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: after remove modal html is like this 
```
<div class="modal-header">
        <h4 class="modal-title" id="myOperationLabel" jhiTranslate="jhipsterSampleApplicationApp.operation.home.createOrEditLabel">Create or edit a Operation</h4>
        <button type="button" class="close" data-dismiss="modal" aria-hidden="true"
                (click)="clear()">&times;</button>
   </div> 
```

and modal will remove

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: all travis build that failed i thin for css and html 
https://travis-ci.org/jhipster/generator-jhipster/jobs/360669286#L5005

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: Well, maybe it's the test that should be changed. 

To me, it does not seem right from a structural point of view, to have a h4 nested under a h1.

In React,  it's also h2 https://github.com/jhipster/generator-jhipster/blob/master/generators/entity-client/templates/react/src/main/webapp/app/entities/entity-detail.tsx.ejs#L65

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: where is this test ? what will be change ? can you help me to change that ?

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: See https://github.com/azizkhani/generator-jhipster/blob/master/generators/entity-client/templates/angular/src/test/javascript/e2e/entities/entity.spec.ts.ejs#L85
These are the templates to generate e2e tests

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: Travis generates about 20 projects then builds them, run unit tests and end to end tests.

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: so either you run these Travis" builds locally, all of them but it takes time or you run only the ones that have failed in real Travis. or you re-generate your sample and run yarn e2e

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: Are you on Windows?

**Comment by gmarziou** [comment]
Also a little advice about git for JHipster pul requests:

- you could squash your commits as one single commit, it's easier to review and history is cleaner
- I found it little harder to merge your changes because you worked on master in your clone, so adding it to mine as new remote to pull changes forced me to create a branch.Usually I just have to pull the branch from remote clone. Next time, please consider working in an issue branch.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs [hunk 2]
```diff
@@ -21,12 +21,10 @@ const keyPrefix = angularAppName + '.'+ entityTranslationKey + '.';
 _%>
 <form name="editForm" role="form" novalidate (ngSubmit)="save()" #editForm="ngForm">
 
-    <div class="modal-header">
-        <h4 class="modal-title" id="my<%= entityClass %>Label" jhiTranslate="<%= keyPrefix %>home.createOrEditLabel">Create or edit a <%= entityClassHumanized %></h4>
-        <button type="button" class="close" data-dismiss="modal" aria-hidden="true"
-                (click)="clear()">&times;</button>
+    <div>
```
Comment (deepu105) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: the classes and filenames needs to be changed to remove dialog from it

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: What is your suggestion for new name ? 
entity-management.component.html
entity-management-detail.component.html
entity-management-delete-dialog.component.html
entity-management-dialog.component.html --> ?

Comment (deepu105) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: entity-management-update.component.html
entity-management-dialog.component.ts

Comment (azizkhani) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: i changed to update but have conflict .how can resolve that?

Comment (gmarziou) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/entity-management-dialog.component.html.ejs]: You pull from our master, merge and resolve locally, then commit and push.

**Comment by pascalgrimaud** [comment]
@gmarziou : maybe it can help. Here my workflow + git config, when I want to test a PR. I used this -> https://gist.github.com/piscisaureus/3342247

Here my `git remote -v` :

```
origin	git@github.com:pascalgrimaud/generator-jhipster.git (fetch)
origin	git@github.com:pascalgrimaud/generator-jhipster.git (push)
upstream	https://github.com/jhipster/generator-jhipster.git (fetch)
upstream	https://github.com/jhipster/generator-jhipster.git (push)
```

Here an extract of `.git/config` :

```
[remote "upstream"]
	url = https://github.com/jhipster/generator-jhipster.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
	fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*
```

So each time I launch `git fetch upstream`, I can do `git checkout pr/XXXX`

**Comment by azizkhani** [comment]
@deepu105 can you help me ? i dont know what is this error
https://travis-ci.org/jhipster/generator-jhipster/jobs/360934987

**Comment by deepu105** [comment]
I'll take a look tomorrow, might be a glitch so first will restart it

**Comment by deepu105** [comment]
Fixed and merged

**Comment by deepu105** [comment]
Thanks @azizkhani, great work

**Comment by gmarziou** [comment]
Congratulations @azizkhani for your first merged PR!

**Comment by jdubois** [comment]
I totally missed this, huge work, congrats too!
BTW you commited the yarn-error.log file by error, I'll remove it

**Comment by jdubois** [comment]
Oh sorry, I don't know how this happened, it's in the diff but it's supposed to be ignored.... forget about my comment, I'm not sure what happened here

**Comment by deepu105** [comment]
I removed it and ignored it as well. Guess you didnt notice my commits on
top of this

Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
Argh there's still an issue here: the links between entities (when there are relationships) are not working anymore!
Otherwise I love this, it's much better than before.
@azizkhani @deepu105 can you have a look, or do you want me to create a specific issue for this?

**Comment by deepu105** [comment]
I'll check, I guess I missed to update the url pattern for it

**Comment by deepu105** [comment]
Its fixed

Thanks & Regards,
Deepu

**Comment by azizkhani** [comment]
I am very happy that could create PR  .I hope can do that again in new issue .

**Comment by gmarziou** [comment]
Well maybe removing modals from user management :)

**Comment by deepu105** [comment]
Its already done :) thanks to me :)

Thanks & Regards,
Deepu

**Comment by piprog** [comment]
Hi Team - Please understand that I love JHipster and use it on current projects. I also haven't seen this full-page edit/new feature (instead of popup modal) yet. And I also appreciate all the hard work it has been to implement this. BUT :) Has there been a discussion of this from a usability/UX perspective? In a to-do demo app with 3-5 functions it may be beautiful (from code-beauty perspective) and consistent to go to a new page all the time, but in real-world scenarios with non-trivial applications such a context switch will trash the user's train of thought and understanding of context (especially if there is an external interruption) and may also remove relevant information from the screen that is needed to complete a task. On the other hand may help with mobile phone support :). - Anyway, I'll check out the new way of operation and will let you know how it goes.

**Comment by gmarziou** [comment]
@piprog your comment would be better located on the issue/proposition rather than on the pull request.
Maybe after you've tried it and also after considering the positive impact on accessibility.

**Comment by ddm7** [comment]
Hi,

It is possible to have both features (modal page and separate page) ?

I'm using JHipster and the application (~100 entities) is in production since 4 months. If you drop modal page I have to refactor all the frontend ... which is not a very simple task. 

I developed also a master-detail page which is using the the "View" feature (View/Edit/Delete). Are you going the keep this feature ?

Thank you

**Comment by ddm7** [comment]
Yes, but I'm loosing all the new features that will come. And JHipster 4 is dead, except bug fixing. 

For example I'm using DTO in my application. In JHipster 4 is in beta. Are you going to make a release for DTO in JHipster 4 ? I think no.

You should take into consideration that some of us are using this framework for production. This is a change with a major impact.

Thank you

**Comment by gmarziou** [comment]
What do you mean by DTO? This exists since JH 3.

JHipster is not a framework, it's a code generator and it makes a huge difference: generated code is your responsibility as if you had written it.
We take into consideration production because each core team member has usually several applications in production.

**Comment by gmarziou** [comment]
Disclaimer: these are my opinions only, maybe other team members have different views.

**Comment by ddm7** [comment]
DTO

WARNING! This is a new feature, of BETA quality.
https://www.jhipster.tech/using-dtos/

**Comment by deepu105** [comment]
@ddm7 we definitely understand your concern but the project has to move forward and Major version are only when we can make breaking changes. Hope you can understand that. We made efforts to make breaking changes as minimal as possible. Its unavoidable for a project that's growing fast. 

Now practically what you could do is below and I personally think it's a good approach to take if you want to upgrade to JHipster 5(I would advise to do it once the stable release is made as the current one is in beta)

1. Upgrade to JHipster 5 globally
2. Create a new folder and copy the `.yo-rc.json` and the `.jhipster` folder from your existing project into it
3. Run `jhipster` to generate the project
4. Use a tool like [Beyond compare](http://scootersoftware.com/) to compare the new project with your existing project and move the changes you want over to the existing project (make sure to be on a new branch)
5. In JHipster 5 the Angular implementation doesn't have many new features, only notable one you might want is the lazy loading of admin modules, and webpack stuff, so just ignore the changes on the edit/delete pages of entities and move the rest and you should be fine

**Comment by gmarziou** [comment]
DTO, this is what I meant, it is in beta since JHipster 3:  https://www.jhipster.tech/2016/08/17/jhipster-release-3.6.0.html

**Comment by ddm7** [comment]
Back to my question. 

If you will release DTO in JHipster 5, the release will apply to JHipster 3 ? Or is for JHipster 5 only ?

**Comment by gmarziou** [comment]
In JH 5 only, but then JHipster is a community project, if someone wants to contribute some changes to a maintenance release, I'm sure we will welcome it.

**Comment by jdubois** [comment]
At some point we should have DTO out of beta - one of the issues was IDE integration, and I *think* I've solved this in JHipster 5. So we could remove the beta tag when JHipster 5 is out.

Concerning maintenance releases, that's a common issue with OSS projects: people work on it for free, and there's no interest to maintain old releases. Companies can of course pay people and/or contribute, and we'll be happy to have that code, but they never do - in fact the problem is with companies that want a free ride, isn't it?

**Comment by ddm7** [comment]
Thank you guys! I'll stop here ... 

Keep up the good work JHipster !
